File: 2024/Cplusplus/day1/day1b.cpp
#include <fstream>
#include <iostream>
#include <vector>

using namespace std;

#define REAL "day1.txt"
#define TEST "test.txt"

int main(void) {
  ifstream file(REAL);

  if (!file) {
    cerr << "Error in Text file\n";
    return 1;
  }

  vector<int> left;
  vector<int> right;

  int left_val;
  int right_val;
  while (file >> left_val >> right_val) {
    left.push_back(left_val);
    right.push_back(right_val);
  }

  // no id is > 100,000
  int r[100000] = {0};

  for (auto &x : right) {
    r[x]++;
  }

  long long total = 0;
  for (auto &i : left) {
    total += i * r[i];
  }

  cout << total << '\n';

  return 0;
}

File: 2024/Typescript/day1b.ts
import * as fs from "fs";

const REAL = "./day1.txt";
// const TEST = "./test.txt";

let content: string;
try {
  content = fs.readFileSync(REAL, "utf-8").trim();
} catch (err) {
  throw new Error(`Falid to parse the file: ${(err as Error).message}`);
}

let left: number[] = [];
let right: number[] = [];

for (let line of content.split("\n")) {
  let [l, r] = line.trim().split(/\s+/);
  left.push(Number(l));
  right.push(Number(r));
}

let freqArr: number[] = new Array(100000).fill(0);

for (let n of right) {
  freqArr[n] += 1;
}

let total = 0;
for (let n of left) {
  total += n * freqArr[n];
}

console.log(total);
